kernel/src/table_changes/log_replay.rs

```diff
@@ -0,0 +1,847 @@
+            const INTEGER: DataType = DataType::INTEGER;
+            const LONG: DataType = DataType::LONG;
+            let string_list: DataType = ArrayType::new(STRING, false).into();
+            let string_string_map = MapType::new(STRING, STRING, false).into();
+            let types_and_names = vec![
+                (STRING, column_name!("add.path")),
+                (STRING, column_name!("remove.path")),
+                (STRING, column_name!("remove.deletionVector.storageType")),
+                (STRING, column_name!("remove.deletionVector.pathOrInlineDv")),
+                (INTEGER, column_name!("remove.deletionVector.offset")),
+                (INTEGER, column_name!("remove.deletionVector.sizeInBytes")),
+                (LONG, column_name!("remove.deletionVector.cardinality")),
+                (STRING, column_name!("cdc.path")),
+                (LONG, column_name!("commitInfo.timestamp")),
+                (STRING, column_name!("metaData.schemaString")),
+                (string_string_map, column_name!("metaData.configuration")),
+                (INTEGER, column_name!("protocol.minReaderVersion")),
+                (INTEGER, column_name!("protocol.minWriterVersion")),
+                (string_list.clone(), column_name!("protocol.readerFeatures")),
+                (string_list, column_name!("protocol.writerFeatures")),
+            ];
+            let (types, names) = types_and_names.into_iter().unzip();
+            (names, types).into()
+        });
+        NAMES_AND_TYPES.as_ref()
+    }
+
+    fn visit<'b>(
+        &mut self,
+        row_count: usize,
+        getters: &[&'b dyn crate::engine_data::GetData<'b>],
+    ) -> DeltaResult<()> {
+        let expected_getters = 15;
+        require!(
+            getters.len() == expected_getters,
+            Error::InternalError(format!(
+                "Wrong number of PreparePhaseVisitor getters: {}",
+                getters.len()
+            ))
+        );
+
+        for i in 0..row_count {
+            if !self.selection_vector[i] {
+                continue;
+            }
+            if let Some(path) = getters[0].get_str(i, "add.path")? {
+                self.add_paths.insert(path.to_string());
+            } else if let Some(path) = getters[1].get_str(i, "remove.path")? {
+                let deletion_vector = visit_deletion_vector_at(i, &getters[2..=6])?;
+                self.scanner
+                    .remove_dvs
+                    .insert(path.to_string(), DvInfo { deletion_vector });
+            } else if getters[7].get_str(i, "cdc.path")?.is_some() {
+                self.scanner.has_cdc_action = true;
+            } else if let Some(timestamp) = getters[8].get_long(i, "commitInfo.timestamp")? {
+                self.scanner.timestamp = timestamp;
+            } else if let Some(schema) = getters[9].get_str(i, "metaData.schemaString")? {
+                let configuration_map_opt: Option<HashMap<_, _>> =
```

Comment:
I suspect we don't need the type annotation. Even if we did, just do `get_map` (strongly typed) instead of `get_opt` (needs type inference).